Database deletion failing with 1.4-25

Failing test - http://uberjenkins.sc.couchbase.com/job/android-listener-client-sync-gateway/65/testReport/junit/testsuites.listener.shared.client_sg/test_changes/test_longpoll_changes_termination_heartbeat/

```
V/Listener( 1677): remoteURL() addr=10.17.1.40
V/Database( 1677): Acme.Utils.ThreadPool(19)-PooledThread: Acme.Serve.Serve$ServeConnection@53593bb4 Committing transaction (level 0)
V/Database( 1677): No pending doc expirations
V/Router  ( 1677): waiting=true & db!=null: call Database.addDatabaseListener()
V/Listener( 1677): CountDownLatch.await() START
V/Router  ( 1677): waiting=true & db!=null: call Database.addDatabaseListener()
V/Listener( 1677): CountDownLatch.await() START
V/Router  ( 1677): waiting=true & db!=null: call Database.addDatabaseListener()
V/Listener( 1677): CountDownLatch.await() START
V/Router  ( 1677): waiting=true & db!=null: call Database.addDatabaseListener()
V/Listener( 1677): CountDownLatch.await() START
V/Router  ( 1677): waiting=true & db!=null: call Database.addDatabaseListener()
V/Listener( 1677): CountDownLatch.await() START
V/Router  ( 1677): waiting=true & db!=null: call Database.addDatabaseListener()
V/Listener( 1677): CountDownLatch.await() START
V/Database( 1677): Acme.Utils.ThreadPool(12)-PooledThread: Acme.Serve.Serve$ServeConnection@535c5b1c Begin transaction (level 0)
E/Database( 1677): Failed to execSQL: CREATE TABLE docs (         doc_id INTEGER PRIMARY KEY,         docid TEXT UNIQUE NOT NULL)
E/Database( 1677): com.couchbase.lite.storage.SQLException: com.couchbase.lite.internal.database.sqlite.exception.SQLiteException: table docs already exists (code 1)
E/Database( 1677): 	at com.couchbase.lite.storage.SQLiteStorageEngineBase.execSQL(SQLiteStorageEngineBase.java:168)
E/Database( 1677): 	at com.couchbase.lite.store.SQLiteStore.runStatements(SQLiteStore.java:2614)
E/Database( 1677): 	at com.couchbase.lite.store.SQLiteStore.initialize(SQLiteStore.java:2624)
E/Database( 1677): 	at com.couchbase.lite.store.SQLiteStore.open(SQLiteStore.java:254)
E/Database( 1677): 	at com.couchbase.lite.Database.open(Database.java:1338)
E/Database( 1677): 	at com.couchbase.lite.Database.open(Database.java:1272)
E/Database( 1677): 	at com.couchbase.lite.router.Router.openDB(Router.java:358)
E/Database( 1677): 	at com.couchbase.lite.router.Router.start(Router.java:477)
E/Database( 1677): 	at com.couchbase.lite.listener.LiteServlet.service(LiteServlet.java:141)
E/Database( 1677): 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:802)
E/Database( 1677): 	at Acme.Serve.Serve$ServeConnection.runServlet(Serve.java:2408)
E/Database( 1677): 	at Acme.Serve.Serve$ServeConnection.parseRequest(Serve.java:2330)
E/Database( 1677): 	at Acme.Serve.Serve$ServeConnection.run(Serve.java:2110)
E/Database( 1677): 	at Acme.Utils$ThreadPool$PooledThread.run(Utils.java:1238)
E/Database( 1677): 	at java.lang.Thread.run(Thread.java:856)
E/Database( 1677): Caused by: com.couchbase.lite.internal.database.sqlite.exception.SQLiteException: table docs already exists (code 1)
E/Database( 1677): 	at com.couchbase.lite.internal.database.sqlite.SQLiteConnection.nativeExecuteForChangedRowCount(Native Method)
E/Database( 1677): 	at com.couchbase.lite.internal.database.sqlite.SQLiteConnection.executeForChangedRowCount(SQLiteConnection.java:617)
E/Database( 1677): 	at com.couchbase.lite.internal.database.sqlite.SQLiteSession.executeForChangedRowCount(SQLiteSession.java:725)
E/Database( 1677): 	at com.couchbase.lite.internal.database.sqlite.SQLiteStatement.executeUpdateDelete(SQLiteStatement.java:67)
E/Database( 1677): 	at com.couchbase.lite.internal.database.sqlite.SQLiteDatabase.executeSql(SQLiteDatabase.java:1660)
E/Database( 1677): 	at com.couchbase.lite.internal.database.sqlite.SQLiteDatabase.execSQL(SQLiteDatabase.java:1589)
E/Database( 1677): 	at com.couchbase.lite.storage.SQLiteStorageEngineBase.execSQL(SQLiteStorageEngineBase.java:166)
E/Database( 1677): 	... 14 more
E/Database( 1677): Cannot initialize database schema
E/Database( 1677): com.couchbase.lite.storage.SQLException: com.couchbase.lite.internal.database.sqlite.exception.SQLiteException: table docs already exists (code 1)
E/Database( 1677): 	at com.couchbase.lite.storage.SQLiteStorageEngineBase.execSQL(SQLiteStorageEngineBase.java:168)
E/Database( 1677): 	at com.couchbase.lite.store.SQLiteStore.runStatements(SQLiteStore.java:2614)
E/Database( 1677): 	at com.couchbase.lite.store.SQLiteStore.initialize(SQLiteStore.java:2624)
E/Database( 1677): 	at com.couchbase.lite.store.SQLiteStore.open(SQLiteStore.java:254)
E/Database( 1677): 	at com.couchbase.lite.Database.open(Database.java:1338)
E/Database( 1677): 	at com.couchbase.lite.Database.open(Database.java:1272)
E/Database( 1677): 	at com.couchbase.lite.router.Router.openDB(Router.java:358)
E/Database( 1677): 	at com.couchbase.lite.router.Router.start(Router.java:477)
E/Database( 1677): 	at com.couchbase.lite.listener.LiteServlet.service(LiteServlet.java:141)
E/Database( 1677): 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:802)
E/Database( 1677): 	at Acme.Serve.Serve$ServeConnection.runServlet(Serve.java:2408)
E/Database( 1677): 	at Acme.Serve.Serve$ServeConnection.parseRequest(Serve.java:2330)
E/Database( 1677): 	at Acme.Serve.Serve$ServeConnection.run(Serve.java:2110)
E/Database( 1677): 	at Acme.Utils$ThreadPool$PooledThread.run(Utils.java:1238)
E/Database( 1677): 	at java.lang.Thread.run(Thread.java:856)
E/Database( 1677): Caused by: com.couchbase.lite.internal.database.sqlite.exception.SQLiteException: table docs already exists (code 1)
E/Database( 1677): 	at com.couchbase.lite.internal.database.sqlite.SQLiteConnection.nativeExecuteForChangedRowCount(Native Method)
E/Database( 1677): 	at com.couchbase.lite.internal.database.sqlite.SQLiteConnection.executeForChangedRowCount(SQLiteConnection.java:617)
E/Database( 1677): 	at com.couchbase.lite.internal.database.sqlite.SQLiteSession.executeForChangedRowCount(SQLiteSession.java:725)
E/Database( 1677): 	at com.couchbase.lite.internal.database.sqlite.SQLiteStatement.executeUpdateDelete(SQLiteStatement.java:67)
E/Database( 1677): 	at com.couchbase.lite.internal.database.sqlite.SQLiteDatabase.executeSql(SQLiteDatabase.java:1660)
E/Database( 1677): 	at com.couchbase.lite.internal.database.sqlite.SQLiteDatabase.execSQL(SQLiteDatabase.java:1589)
E/Database( 1677): 	at com.couchbase.lite.storage.SQLiteStorageEngineBase.execSQL(SQLiteStorageEngineBase.java:166)
E/Database( 1677): 	... 14 more
V/Database( 1677): Acme.Utils.ThreadPool(12)-PooledThread: Acme.Serve.Serve$ServeConnection@535c5b1c CANCEL transaction (level 0)
W/System.err( 1677): [Thu Jan 19 01:53:54 GMT 2017] TJWS: Unexpected problem running servlet
W/System.err( 1677): java.lang.NullPointerException
W/System.err( 1677): 	at com.couchbase.lite.store.SQLiteStore.endTransaction(SQLiteStore.java:2417)
W/System.err( 1677): 	at com.couchbase.lite.store.SQLiteStore.open(SQLiteStore.java:313)
W/System.err( 1677): 	at com.couchbase.lite.Database.open(Database.java:1338)
W/System.err( 1677): 	at com.couchbase.lite.Database.open(Database.java:1272)
W/System.err( 1677): 	at com.couchbase.lite.router.Router.openDB(Router.java:358)
W/System.err( 1677): 	at com.couchbase.lite.router.Router.start(Router.java:477)
W/System.err( 1677): 	at com.couchbase.lite.listener.LiteServlet.service(LiteServlet.java:141)
W/System.err( 1677): 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:802)
W/System.err( 1677): 	at Acme.Serve.Serve$ServeConnection.runServlet(Serve.java:2408)
W/System.err( 1677): 	at Acme.Serve.Serve$ServeConnection.parseRequest(Serve.java:2330)
W/System.err( 1677): 	at Acme.Serve.Serve$ServeConnection.run(Serve.java:2110)
W/System.err( 1677): 	at Acme.Utils$ThreadPool$PooledThread.run(Utils.java:1238)
W/System.err( 1677): 	at java.lang.Thread.run(Thread.java:856)
W/System.err( 1677): [Thu Jan 19 01:53:54 GMT 2017] TJWS: Unexpected problem running servlet: java.lang.NullPointerException
D/dalvikvm( 1677): GC_FOR_ALLOC freed 503K, 13% free 8196K/9415K, paused 3ms, total 3ms
D/dalvikvm( 1677): GC_CONCURRENT freed 22K, 14% free 8174K/9415K, paused 12ms+0ms, total 14ms
V/Router  ( 1677): waiting=true & db!=null: call Database.addDatabaseListener()
V/Listener( 1677): CountDownLatch.await() START
V/Router  ( 1677): waiting=true & db!=null: call Database.addDatabaseListener()
V/Listener( 1677): CountDownLatch.await() START
V/Router  ( 1677): waiting=true & db!=null: call Database.addDatabaseListener()
V/Listener( 1677): CountDownLatch.await() START
V/Router  ( 1677): waiting=true & db!=null: call Database.addDatabaseListener()
V/Listener( 1677): CountDownLatch.await() START
V/Router  ( 1677): waiting=true & db!=null: call Database.addDatabaseListener()
V/Listener( 1677): CountDownLatch.await() START
V/Router  ( 1677): waiting=true & db!=null: call Database.addDatabaseListener()
V/Listener( 1677): CountDownLatch.await() START
```

Full logcat - http://uberjenkins.sc.couchbase.com/job/android-listener-client-sync-gateway/65/artifact/results/logs/LiteServAndroid-test_longpoll_changes_termination_heartbeat-2017-01-18%2017%3A53%3A51.401301.txt

--- a/src/main/java/com/couchbase/lite/Database.java
+++ b/src/main/java/com/couchbase/lite/Database.java
@@ -1457,73 +1457,78 @@ public class Database implements StoreDelegate {
     @InterfaceAudience.Public
     public synchronized boolean close() {
         storeRef.await();
-        try {
-            closing.set(true);
 
-            if (!open.get()) {
-                // Ensure that the database is forgotten:
-                manager.forgetDatabase(this);
-                return false;
-            }
+        // NOTE: synchronized Manager.lockDatabases to prevent Manager to give the deleting Database
+        //       instance. See also `Manager.getDatabase(String, boolean)`.
+        synchronized (manager.lockDatabases) {
+            try {
+                closing.set(true);
 
-            synchronized (databaseListeners) {
-                for (DatabaseListener listener : databaseListeners)
-                    listener.databaseClosing();
-            }
-
-            synchronized (lockViews) {
-                if (views != null) {
-                    for (View view : views.values())
-                        view.close();
-                }
-                views = null;
-            }
-
-            // Make all replicators stop and wait:
-            boolean stopping = false;
-            synchronized (activeReplicators) {
-                for (Replication replicator : activeReplicators) {
-                    if (replicator.getStatus() == Replication.ReplicationStatus.REPLICATION_STOPPED)
-                        continue;
-                    replicator.stop();
-                    stopping = true;
+                if (!open.get()) {
+                    // Ensure that the database is forgotten:
+                    manager.forgetDatabase(this);
+                    return false;
                 }
 
-                // maximum wait time per replicator is 60 sec.
-                // total maximum wait time for all replicators is between 60sec and 119 sec.
-                long timeout = Replication.DEFAULT_MAX_TIMEOUT_FOR_SHUTDOWN * 1000;
-                long startTime = System.currentTimeMillis();
-                while (activeReplicators.size() > 0 && stopping &&
-                        (System.currentTimeMillis() - startTime) < timeout) {
-                    try {
-                        activeReplicators.wait(timeout);
-                    } catch (InterruptedException e) {
+                synchronized (databaseListeners) {
+                    for (DatabaseListener listener : databaseListeners)
+                        listener.databaseClosing();
+                }
+
+                synchronized (lockViews) {
+                    if (views != null) {
+                        for (View view : views.values())
+                            view.close();
                     }
+                    views = null;
                 }
-                // clear active replicators:
-                activeReplicators.clear();
+
+                // Make all replicators stop and wait:
+                boolean stopping = false;
+                synchronized (activeReplicators) {
+                    for (Replication replicator : activeReplicators) {
+                        if (replicator.getStatus() == Replication.ReplicationStatus.REPLICATION_STOPPED)
+                            continue;
+                        replicator.stop();
+                        stopping = true;
+                    }
+
+                    // maximum wait time per replicator is 60 sec.
+                    // total maximum wait time for all replicators is between 60sec and 119 sec.
+                    long timeout = Replication.DEFAULT_MAX_TIMEOUT_FOR_SHUTDOWN * 1000;
+                    long startTime = System.currentTimeMillis();
+                    while (activeReplicators.size() > 0 && stopping &&
+                            (System.currentTimeMillis() - startTime) < timeout) {
+                        try {
+                            activeReplicators.wait(timeout);
+                        } catch (InterruptedException e) {
+                        }
+                    }
+                    // clear active replicators:
+                    activeReplicators.clear();
+                }
+
+                // Clear all replicators:
+                allReplicators.clear();
+
+                // cancel purge timer
+                cancelPurgeTimer();
+
+                // Close Store:
+                if (store != null)
+                    store.close();
+
+                // Clear document cache:
+                clearDocumentCache();
+
+                // Forget database:
+                manager.forgetDatabase(this);
+
+                open.set(false);
+                return true;
+            } finally {
+                closing.set(false);
             }
-
-            // Clear all replicators:
-            allReplicators.clear();
-
-            // cancel purge timer
-            cancelPurgeTimer();
-
-            // Close Store:
-            if (store != null)
-                store.close();
-
-            // Clear document cache:
-            clearDocumentCache();
-
-            // Forget database:
-            manager.forgetDatabase(this);
-
-            open.set(false);
-            return true;
-        } finally {
-            closing.set(false);
         }
     }
 

--- a/src/main/java/com/couchbase/lite/Database.java
+++ b/src/main/java/com/couchbase/lite/Database.java
@@ -394,23 +394,30 @@ public class Database implements StoreDelegate {
      * Deletes the database.
      */
     @InterfaceAudience.Public
-    public void delete() throws CouchbaseLiteException {
-        if (open.get()) {
-            if (!close()) {
-                throw new CouchbaseLiteException("The database was open, and could not be closed",
-                        Status.INTERNAL_SERVER_ERROR);
+    public synchronized void delete() throws CouchbaseLiteException {
+        // NOTE: synchronized Manager.lockDatabases to prevent Manager to give the deleting Database
+        //       instance. See also `Manager.getDatabase(String, boolean)`.
+        synchronized (manager.lockDatabases) {
+            Log.v(TAG, "Deleting %s", this);
+
+            if (open.get()) {
+                if (!close())
+                    throw new CouchbaseLiteException("The database was open, and could not be closed",
+                            Status.INTERNAL_SERVER_ERROR);
             }
-        }
-        manager.forgetDatabase(this);
-        if (!exists()) {
-            return;
-        }
 
-        File dir = new File(path);
-        if (!FileDirUtils.deleteRecursive(dir))
-            throw new CouchbaseLiteException("Was not able to delete the database directory",
-                    Status.INTERNAL_SERVER_ERROR);
+            manager.forgetDatabase(this);
 
+            if (!exists())
+                return;
+
+            File dir = new File(path);
+            if (!FileDirUtils.deleteRecursive(dir))
+                throw new CouchbaseLiteException("Was not able to delete the database directory",
+                        Status.INTERNAL_SERVER_ERROR);
+
+            Log.v(TAG, "Deleted %s", this);
+        }
     }
 
     /**
@@ -1232,7 +1239,7 @@ public class Database implements StoreDelegate {
     }
 
     @InterfaceAudience.Private
-    public boolean exists() {
+    public synchronized boolean exists() {
         return new File(path).exists();
     }
 
@@ -1268,7 +1275,7 @@ public class Database implements StoreDelegate {
     }
 
     @InterfaceAudience.Private
-    public synchronized void open() throws CouchbaseLiteException {
+    public void open() throws CouchbaseLiteException {
         open(manager.getDefaultOptions(name));
     }
 

--- a/src/main/java/com/couchbase/lite/Manager.java
+++ b/src/main/java/com/couchbase/lite/Manager.java
@@ -77,6 +77,7 @@ public final class Manager {
 
     private ManagerOptions options;
     private File directoryFile;
+    /*package*/ Object lockDatabases = new Object();
     private Map<String, Database> databases;
     private Map<String, Object> encryptionKeys;
     private List<Replication> replications;
@@ -262,24 +263,26 @@ public final class Manager {
      */
     @InterfaceAudience.Public
     public void close() {
-        Log.d(Database.TAG, "Closing " + this);
-        // Close all database:
-        // Snapshot of the current open database to avoid concurrent modification as
-        // the database will be forgotten (removed from the databases map) when it is closed:
-        Database[] openDbs = databases.values().toArray(new Database[databases.size()]);
-        for (Database database : openDbs) {
-            database.close();
-        }
-        databases.clear();
+        synchronized (lockDatabases) {
+            Log.d(Database.TAG, "Closing " + this);
 
-        // Stop reachability:
-        context.getNetworkReachabilityManager().stopListening();
+            // Close all database:
+            // Snapshot of the current open database to avoid concurrent modification as
+            // the database will be forgotten (removed from the databases map) when it is closed:
+            Database[] openDbs = databases.values().toArray(new Database[databases.size()]);
+            for (Database database : openDbs)
+                database.close();
+            databases.clear();
 
-        // Shutdown ScheduledExecutorService:
-        if (workExecutor != null && !workExecutor.isShutdown()) {
-            Utils.shutdownAndAwaitTermination(workExecutor);
+            // Stop reachability:
+            context.getNetworkReachabilityManager().stopListening();
+
+            // Shutdown ScheduledExecutorService:
+            if (workExecutor != null && !workExecutor.isShutdown())
+                Utils.shutdownAndAwaitTermination(workExecutor);
+
+            Log.d(Database.TAG, "Closed " + this);
         }
-        Log.d(Database.TAG, "Closed " + this);
     }
 
     /**
@@ -512,7 +515,9 @@ public final class Manager {
      */
     @InterfaceAudience.Private
     public Collection<Database> allOpenDatabases() {
-        return databases.values();
+        synchronized (lockDatabases) {
+            return databases.values();
+        }
     }
 
     /**
@@ -551,25 +556,28 @@ public final class Manager {
      * @exclude
      */
     @InterfaceAudience.Private
-    public synchronized Database getDatabase(String name, boolean mustExist) {
-        if (options.isReadOnly())
-            mustExist = true;
-        Database db = databases.get(name);
-        if (db == null) {
-            if (!isValidDatabaseName(name))
-                throw new IllegalArgumentException("Invalid database name: " + name);
-            String path = pathForDatabaseNamed(name);
-            if (path == null)
-                return null;
-            db = new Database(path, name, this, options.isReadOnly());
-            if (mustExist && !db.exists()) {
-                Log.i(Database.TAG, "mustExist is true and db (%s) does not exist", name);
-                return null;
+    public Database getDatabase(String name, boolean mustExist) {
+        synchronized (lockDatabases) {
+            if (options.isReadOnly())
+                mustExist = true;
+            Database db = databases.get(name);
+            if (db == null) {
+                if (!isValidDatabaseName(name))
+                    throw new IllegalArgumentException("Invalid database name: " + name);
+                String path = pathForDatabaseNamed(name);
+                if (path == null)
+                    return null;
+                db = new Database(path, name, this, options.isReadOnly());
+                if (mustExist && !db.exists()) {
+                    Log.i(Database.TAG, "mustExist is true and db (%s) does not exist", name);
+                    return null;
+                }
+                db.setName(name);
+                databases.put(name, db);
             }
-            db.setName(name);
-            databases.put(name, db);
+            Log.v(Log.TAG_DATABASE, "getDatabase() %s %s", this, db);
+            return db;
         }
-        return db;
     }
 
     /**
@@ -1003,21 +1011,27 @@ public final class Manager {
      */
     @InterfaceAudience.Private
     protected void forgetDatabase(Database db) {
-        // remove from cached list of dbs
-        databases.remove(db.getName());
+        synchronized (lockDatabases) {
+            Log.v(Log.TAG_DATABASE, "Fogetting forgetDatabase() %s %s", this, db);
 
-        // remove from list of replications
-        // TODO: should there be something that actually stops the replication(s) first?
-        Iterator<Replication> replicationIterator = this.replications.iterator();
-        while (replicationIterator.hasNext()) {
-            Replication replication = replicationIterator.next();
-            if (replication.getLocalDatabase().getName().equals(db.getName())) {
-                replicationIterator.remove();
+            // remove from cached list of dbs
+            databases.remove(db.getName());
+
+            // remove from list of replications
+            // TODO: should there be something that actually stops the replication(s) first?
+            Iterator<Replication> replicationIterator = this.replications.iterator();
+            while (replicationIterator.hasNext()) {
+                Replication replication = replicationIterator.next();
+                if (replication.getLocalDatabase().getName().equals(db.getName())) {
+                    replicationIterator.remove();
+                }
             }
-        }
 
-        // Remove registered encryption key if available:
-        encryptionKeys.remove(db.getName());
+            // Remove registered encryption key if available:
+            encryptionKeys.remove(db.getName());
+
+            Log.v(Log.TAG_DATABASE, "Forgot forgetDatabase() %s %s", this, db);
+        }
     }
 
     /**

--- a/src/main/java/com/couchbase/lite/replicator/ReplicationInternal.java
+++ b/src/main/java/com/couchbase/lite/replicator/ReplicationInternal.java
@@ -878,17 +878,7 @@ abstract class ReplicationInternal implements BlockingQueueListener {
                         Map<String, Object> response = (Map<String, Object>) result;
                         body.put("_rev", response.get("rev"));
                         remoteCheckpoint = body;
-                        boolean isOpen = false;
-                        try {
-                            if (db != null) {
-                                db.open();
-                                isOpen = true;
-                            }
-                        } catch (CouchbaseLiteException ex) {
-                            Log.w(Log.TAG_SYNC, "%s: Cannot open the database", ex, this);
-                        }
-
-                        if (isOpen) {
+                        if (db != null && db.isOpen()) {
                             Log.d(Log.TAG_SYNC,
                                     "%s: saved remote checkpoint, updating local checkpoint. RemoteCheckpoint: %s",
                                     this, remoteCheckpoint);

--- a/src/main/java/com/couchbase/lite/router/Router.java
+++ b/src/main/java/com/couchbase/lite/router/Router.java
@@ -622,6 +622,7 @@ public class Router implements Database.ChangeListener, Database.DatabaseListene
                 connection.getResponseBody() == null &&
                 connection.getHeaderField("Content-Type") == null &&
                 dontOverwriteBody == false) {
+                connection.getResHeader().add("Content-Type", CONTENT_TYPE_JSON);
                 connection.setResponseBody(new Body("{\"ok\":true}".getBytes()));
             }
 
@@ -1141,13 +1142,17 @@ public class Router implements Database.ChangeListener, Database.DatabaseListene
         return new Status(Status.CREATED);
     }
 
-    public Status do_DELETE_Database(Database _db, String _docID, String _attachmentName)
-            throws CouchbaseLiteException {
+    public Status do_DELETE_Database(Database _db, String _docID, String _attachmentName) {
         if (getQuery("rev") != null) {
             return new Status(Status.BAD_REQUEST);
             // CouchDB checks for this; probably meant to be a document deletion
         }
-        db.delete();
+        try {
+            db.delete();
+        } catch (CouchbaseLiteException e) {
+            Log.w(TAG, "Failed to delete database: do_DELETE_Database()", e);
+            return e.getCBLStatus();
+        }
         return new Status(Status.OK);
     }
 

